iptables: add support for NOTRACK rules for pod to pod traffic

jibi opened the pull request:
When running in KPR mode, all pod-to-pod traffic goes through connection
tracking twice: the first time traffic is tracked by netfilter and the
second one by the BPF datapath.

This is suboptimal as doing conntrack at the netfilter layer is not
strictly required for the operation of the CNI, but it has some
important performance implications.

This commit introduces a new agent option,
--install-no-conntrack-iptables-rules (enabled by default) which
installs `-j NOTRACK` Iptables rules that match all traffic from and to
the native-routing-cidr CIDR, to prevent netfilter from tracking all
pod-to-pod traffic.

This feature can be only enabled when Cilium is:

* running in direct routing
* running in full KPR mode
* not running its CNI chained on top of other CNI plugins
* not running in a managed Kubernetes environment such as AKS, EKS or GKE

There are a few reasons for these specific requirements:

* when running in tunneling mode (and with full KPR), the pod-to-pod
  traffic is encapsulated and decapsulated in the context of the BPF
  datapath program and so it is already skipping the netfilter raw table
  where connection tracking happens.
* when running in kube-proxy mode conntrack is required to perform NAT
* we cannot make assumptions on how a generic CNI is going to use
  netfilter conntrack

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

qmonnet commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1226,6 +1247,14 @@ func (m *IptablesManager) InstallRules(ifName string) error {
+		podsCIDR := option.Config.IPv4NativeRoutingCIDR().String()
+
+		if err := m.addNoTrackPodToPodRules("iptables", podsCIDR); err != nil {
+			return fmt.Errorf("cannot install: %s", err)
```
Nit: Can we expand this error message a little bit?

jibi replied:
I'll change it to something like `cannot install no pod to pod CT rules` to keep it consistent with the other errors :+1:

qmonnet replied:
Thanks! The `cannot install no` reads a little strange, maybe `cannot install rules to skip pod to pod CT`? Or with quotes, `cannot install 'no pod to pod CT' rules`.

qmonnet commented on Documentation/operations/upgrade.rst:
```diff
@@ -366,6 +366,11 @@ New Options
   and has the same desired effect as ``masquerade`` option.
 * ``cni.exclusive``: Use to toggle Cilium installing itself as the only available CNI
   plugin on all nodes.
+* ``install-no-conntrack-iptables-rules``: This option, by default set to true,
```
This is quite a long option name, what about `install-notrack-rules`?

jibi replied:
No strong opinions here, but I feel like we are loosing some context with your suggestion

qmonnet replied:
But then documentation should help and bring this context? :shrug: No strong opinion either, I was trying to have it shorter to type but keep the current version if you feel this is more correct.

qmonnet commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -125,6 +125,21 @@ func getFeedRule(name, args string) []string {
+// routing mode (as in tunneling mode the encapsulated traffic is already
+// skipping netfilter conntrack).
+func skipPodToPodConntrack(ipv6 bool) bool {
+	return !ipv6 && option.Config.EnableIPv4 &&
```
Why not IPv6?

jibi replied:
Because the native routing CIDR (which is used to select all the pod traffic) can only be an IPv4 CIDR.
The `ipv6` check is needed as this function is also [called from](https://github.com/cilium/cilium/pull/15264/files#diff-465a6b3f4ab9e1e94fe39e5aefe81c41eaa1dc99c66500161072e564029c2f51R756) `InstallEndpointNoTrackRules`, which is a different thing (it installs NOCT rules for specific endpoints rather than a global rule to match all pod to pod traffic).

So, if we are trying to install a per endpoint NOCT rule:
* if the rule is for IPv4 and the pod to pod NOCT option is enabled, we skip installing more specific per endpoint rules
* otherwise, for IPv6, we install it

qmonnet replied:
Ok, thanks for the detailed explanation. Would that be worth mentioning in the comments above the function?

qmonnet reviewed (changes requested):
Looks good! Some remarks below.

Would this need a CI test (thinking of DatapathConfiguration.go) to check that the rules are correctly installed, and that no conntrack entry are created for pod-to-pod traffic?

kkourt commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -744,15 +764,15 @@ func InstallNoTrackRules(IP string, port uint16, ipv6 bool) error {
 			return err
 		}
-		if err := noTrackRules(prog, "-A", IP, p, false); err != nil {
+		if err := endpointNoTrackRules(prog, "-A", IP, p, false); err != nil {
```
This is outside of the scope of this PR, but I wonder if we need to call `RemoveEndpointNoTrackRules` (or do something similar) to remove the previous rule that was successfully inserted here.

jibi replied:
Unless I'm missing something I don't think it's possible to have per-endpoint no CT rules in place when the `--install-no-conntrack-iptables-rules` option is set :thinking: (as it's not something you can change after starting the agent)

kkourt replied:
Ah -- I meant if a previous call (e.g., the one on line 767) succeeds but this fails.

kkourt reviewed (approved):
Thanks for the well-documented PR!

qmonnet commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1226,6 +1257,14 @@ func (m *IptablesManager) InstallRules(ifName string) error {
+		podsCIDR := option.Config.IPv4NativeRoutingCIDR().String()
+
+		if err := m.addNoTrackPodToPodRules("iptables", podsCIDR); err != nil {
+			return fmt.Errorf("Cannot install rules to skip pod to pod CT: %s", err)
```
For error messages, use `%w`. The detailed motivation was in [this post](https://blog.golang.org/go1.13-errors). Sorry I missed this on the previous review :/.
```suggestion
			return fmt.Errorf("Cannot install rules to skip pod to pod CT: %w", err)
```

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -125,6 +125,24 @@ func getFeedRule(name, args string) []string {
 	return append(argsList, ruleTail...)
```
Nit: The message for the first commit seems outdated compared to the PR description. It's missing a third point for managed Kubernetes.

jibi replied:
It's part of the second commit description, but maybe I should just merge these 2 commits

qmonnet reviewed (approved):
Thank you!

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -764,7 +788,13 @@ func InstallNoTrackRules(IP string, port uint16, ipv6 bool) error {
+	// skipping conntrack for all pod to pod traffic.
+	if skipPodToPodConntrack(ipv6) {
+		return nil
+	}
```
Why?

What happens if someone:

1. creates a pod with `io.cilium.no-track-port`;
1. enables `install-no-conntrack-iptables-rules` and restart the agent;
2. remove annotation `io.cilium.no-track-port` from the pod;
3. disables `install-no-conntrack-iptables-rules`.

Won't there be a NOTRACK rule for the pod even though the annotation is gone?

jibi replied:
Ah, good point.
I thought on stop (or restart) the agent would clean all the Iptables rules it installed, but I just checked and that's not the case, so I need to rethink this :+1:

jibi replied:
Update: I think we are both right: old rules for the `io.cilium.no-track-port` annotation _should_ get removed (and then eventually reinstalled) during the agent initialization. The problem is that these rules are not being added to the Cilium managed chains (e.g. `CILIUM_PRE_raw`, `CILIUM_OUT_raw`) but rather to the default Iptables chains.

As a test (this is Cilium from master) I deployed a pod with the annotation and the rules show up correctly:
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L PREROUTING
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
CILIUM_PRE_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_PRE_raw */
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L OUTPUT
Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_OUTPUT_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT_raw */
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         tcp  --  10.11.0.14           anywhere             tcp spt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
CT         udp  --  10.11.0.14           anywhere             udp spt:domain NOTRACK
```

but just by restarting Cilium you get duplicate rules:
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L PREROUTING
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
CILIUM_PRE_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_PRE_raw */
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L OUTPUT
Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_OUTPUT_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT_raw */
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         tcp  --  10.11.0.14           anywhere             tcp spt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
CT         udp  --  10.11.0.14           anywhere             udp spt:domain NOTRACK
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         tcp  --  10.11.0.14           anywhere             tcp spt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
CT         udp  --  10.11.0.14           anywhere             udp spt:domain NOTRACK
```

I filled a PR to fix this: https://github.com/cilium/cilium/pull/15411

And this fixes also your original case:

Cilium running with `install-no-conntrack-iptables-rules=false` and a pod with the `io.cilium.no-track-port: "53"`:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
CILIUM_PRE_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_PRE_raw */

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_OUTPUT_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT_raw */

Chain CILIUM_OUTPUT_raw (1 references)
target     prot opt source               destination
CT         all  --  anywhere             anywhere             mark match 0xa00/0xfffffeff /* cilium: NOTRACK for proxy return traffic */ NOTRACK
CT         all  --  anywhere             anywhere             mark match 0xa00/0xfffffeff /* cilium: NOTRACK for proxy return traffic */ NOTRACK
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         tcp  --  10.11.0.14           anywhere             tcp spt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
CT         udp  --  10.11.0.14           anywhere             udp spt:domain NOTRACK

Chain CILIUM_PRE_raw (1 references)
target     prot opt source               destination
CT         all  --  anywhere             anywhere             mark match 0x200/0xf00 /* cilium: NOTRACK for proxy traffic */ NOTRACK
CT         tcp  --  anywhere             10.11.0.14           tcp dpt:domain NOTRACK
CT         udp  --  anywhere             10.11.0.14           udp dpt:domain NOTRACK
```

The I enable `install-no-conntrack-iptables-rules` and restart the agent (the annotation is still there):

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium/mydev$ sudo iptables -t raw -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination
CILIUM_PRE_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_PRE_raw */

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_OUTPUT_raw  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT_raw */

Chain CILIUM_OUTPUT_raw (1 references)
target     prot opt source               destination
CT         all  --  10.0.0.0/16          10.0.0.0/16          /* cilium: NOTRACK for pod-to-pod traffic */ NOTRACK
CT         all  --  anywhere             anywhere             mark match 0xa00/0xfffffeff /* cilium: NOTRACK for proxy return traffic */ NOTRACK
CT         all  --  anywhere             anywhere             mark match 0xa00/0xfffffeff /* cilium: NOTRACK for proxy return traffic */ NOTRACK

Chain CILIUM_PRE_raw (1 references)
target     prot opt source               destination
CT         all  --  10.0.0.0/16          10.0.0.0/16          /* cilium: NOTRACK for pod-to-pod traffic */ NOTRACK
CT         all  --  anywhere             anywhere             mark match 0x200/0xf00 /* cilium: NOTRACK for proxy traffic */ NOTRACK
```

pchaigno replied:
> The I enable `install-no-conntrack-iptables-rules` and restart the agent (the annotation is still there):

You mean `the annotation is not there anymore`, right?

jibi replied:
What I meant was: I didn't edit the pod manifest so the `io.cilium.no-track-port` annotation was still there:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ k get pod ubuntu -o yaml | grep 'io.cilium.no-track-port' -B 2
metadata:
  annotations:
    io.cilium.no-track-port: "53"
```

but after enabling the `install-no-conntrack-iptables-rules` option and restarting the agent the per-endpoint rules were replaced with the  `CT         all  --  10.0.0.0/16          10.0.0.0/16          /* cilium: NOTRACK for pod-to-pod traffic */ NOTRACK
` ones

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1325,6 +1363,33 @@ func (m *IptablesManager) addCiliumNoTrackXfrmRules() error {
+		"-I", ciliumPreRawChain,
+		"-s", podsCIDR,
+		"-d", podsCIDR,
+		"-m", "comment", "--comment", "cilium: NOTRACK for pod to pod traffic",
```
Nit: `pod-to-pod traffic`.

pchaigno commented on pkg/option/config.go:
```diff
@@ -3201,6 +3201,22 @@ func (c *DaemonConfig) calculateDynamicBPFMapSizes(totalMemory uint64, dynamicSi
+// The extra logic to check that all the individual features are enabled is
+// required to deal with the case when KubeProxyReplacement mode is set to
+// "probe" (as Cilium may or may not be running in strict mode).
+func (c *DaemonConfig) KubeProxyReplacementIsRunningStrict() bool {
```
Calling this `strict` feels a bit incorrect. Maybe we should call this `KubeProxyReplacementFullyEnabled` or similar.

pchaigno commented on Documentation/operations/upgrade.rst:
```diff
@@ -366,6 +366,11 @@ New Options
+  can be used to control if Cilium should install some extra Iptables rules to
+  skip doing netfilter connection tracking on all pod to pod traffic. Disabling
+  connection tracking is only possible when Cilium is running in direct routing
+  mode and in a kube-proxy free configuration.
```
I'm not sure if we've been careful when saying "kube-proxy free" so far in the docs, but maybe we should instead say `and using the kube-proxy replacement`? "kube-proxy free" suggest that kube-proxy is not running.

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -138,6 +138,7 @@ func getFeedRule(name, args string) []string {
 // at the moment.
 func skipPodToPodConntrack(ipv6 bool) bool {
 	return !ipv6 && option.Config.EnableIPv4 &&
+		option.Config.InstallNoConntrackIptRules &&
```
Could we detect all of the other conditions below during startup and, if any is not met, disable `InstallNoConntrackIptRules` with an info message? That's how we currently treat other, similar options such as BPF host routing.

Once we have that, we only need to check for `InstallNoConntrackIptRules` here.

pchaigno reviewed (changes requested):
Couple of nits below. The main point is probably the interaction with the `io.cilium.no-track-port` annotation.

By the looks of it, this pull request will also require changes in `cilium-cli`.

pchaigno commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -125,6 +125,25 @@ func getFeedRule(name, args string) []string {
+		option.Config.InstallNoConntrackIptRules &&
+		option.Config.Tunnel == option.TunnelDisabled &&
+		option.Config.IPv4NativeRoutingCIDR() != nil &&
+		option.Config.KubeProxyReplacementIsRunningStrict()
```
We may also need to disable if CNI chaining is enabled. I'm not too sure how that would play with e.g. Azure CNI and AWS-CNI.

jibi replied:
OK, I'm also convinced that we should disable it (at least for now) when CNI chaining is enabled: looking at the docs there are [quite a few different supported configurations](https://docs.cilium.io/en/v1.9/gettingstarted/cni-chaining/) and each of them would require some more thinking to determine if we can relax this constraint.

I followed the same approach I used to the managed Kubernetes environment and disable it in the helm config-map template: https://github.com/cilium/cilium/blob/99ff3b55e2f68277236b703866a026e6a2f17824/install/kubernetes/cilium/templates/cilium-configmap.yaml#L440

pchaigno commented on Documentation/operations/upgrade.rst:
```diff
@@ -366,6 +366,11 @@ New Options
   plugin on all nodes.
+* ``install-no-conntrack-iptables-rules``: This option, by default set to true,
+  can be used to control if Cilium should install some extra Iptables rules to
+  skip doing netfilter connection tracking on all pod-to-pod traffic. Disabling
```
```suggestion
  skip netfilter connection tracking on all pod-to-pod traffic. Disabling
```

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// already skipping netfilter conntrack.
+		if option.Config.Tunnel != option.TunnelDisabled {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires the agent to run in direct routing mode. Disabling it", option.InstallNoConntrackIptRules)
```
```suggestion
			log.Infof("%s requires the agent to run in direct routing mode. Disabling it.", option.InstallNoConntrackIptRules)
```

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// traffic, can only be an IPv4 CIDR at the moment.
+		if !option.Config.EnableIPv4 {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires IPv4 support. Disabling it", option.InstallNoConntrackIptRules)
```
```suggestion
			log.Infof("%s requires IPv4 support. Disabling it.", option.InstallNoConntrackIptRules)
```

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -334,6 +334,16 @@ func initKubeProxyReplacementOptions() (strict bool) {
+		// required for NAT operations
+		if !option.Config.KubeProxyReplacementFullyEnabled() {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires the agent to run in full KPR mode. Disabling it", option.InstallNoConntrackIptRules)
```
```suggestion
			log.Infof("%s requires the agent to run with %s=%s. Disabling it.", option.InstallNoConntrackIptRules, option.KubeProxyReplacement, option.KubeProxyReplacementStrict)
```

jibi replied:
I don't think this is strictly correct as even with `--kube-proxy-replacement probe` it's possible to enable `--install-no-conntrack-iptables-rules` if all the KPR features get enabled

pchaigno replied:
Ah, right. I'm a bit worried the current version of the error message isn't actionable by users though. Maybe we should advice to run with `kube-proxy-replacement=strict` without stating the feature *requires* that? It's probably best if users move from `probe` to `strict` anyway if they can.

jibi replied:
Makes sense :+1: Let's stick with your original suggestion

pchaigno reviewed (approved):
Only nits below. LGTM!

nathanjsweet commented on Documentation/operations/upgrade.rst:
```diff
@@ -366,6 +366,11 @@ New Options
 * ``cni.exclusive``: Use to toggle Cilium installing itself as the only available CNI
   plugin on all nodes.
+* ``install-no-conntrack-iptables-rules``: This option, by default set to true,
+  can be used to control if Cilium should install some extra Iptables rules to
```
nit: The copy is a bit awkward. 
```suggestion
installs some extra Iptables rules to
```

nathanjsweet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// already skipping netfilter conntrack.
+		if option.Config.Tunnel != option.TunnelDisabled {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires the agent to run in direct routing mode. Disabling it.", option.InstallNoConntrackIptRules)
```
This should be a warning. An option being explicitly ignored should be a notable event.

nathanjsweet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// traffic, can only be an IPv4 CIDR at the moment.
+		if !option.Config.EnableIPv4 {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires IPv4 support. Disabling it.", option.InstallNoConntrackIptRules)
```
Warning.

nathanjsweet commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -334,6 +334,17 @@ func initKubeProxyReplacementOptions() (strict bool) {
+		// required for NAT operations
+		if !option.Config.KubeProxyReplacementFullyEnabled() {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Infof("%s requires the agent to run with %s=%s. Disabling it.",
```
Warning.

nathanjsweet reviewed (changes requested):
Some nits.

tgraf reviewed (changes requested):
I don't understand the requirements on not running in managed k8s.

EKS: Native ENI (direct routing) + KPR in probe. So we should probably enable if KPR gets enabled?
GKE: Native routing + k8s IPAM + KPR
AKS: Azure IPAM (direct routing) + KPR

So assuming that the kernel image is recent enough for KPR, we should be able to enable it on all of them?

tgraf commented on pkg/defaults/defaults.go:
```diff
@@ -408,4 +408,7 @@ const (
 	EnableBPFBypassFIBLookup = true
+
+	// InstallNoConntrackRules instructs Cilium to install Iptables rules to skip netfilter connection tracking on all pod-to-pod traffic.
+	InstallNoConntrackIptRules = true
```
Given the discussion on Slack, I can't see how we can safely enable this by default without risking at least some users just to have non-functional clusters with hard to debug problems. Am I missing something?

The default should be guaranteed to succeed on all platforms.

jibi replied:
I'll change the default value to false :+1:

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// already skipping netfilter conntrack.
+		if option.Config.Tunnel != option.TunnelDisabled {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Warnf("%s requires the agent to run in direct routing mode. Disabling it.", option.InstallNoConntrackIptRules)
```
Given `InstallNoConntrackIptRules` now defaults to false, the warnings need to be fatals. The user explicitly asked for something and we refused to enable, so we should be loud about it.

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// already skipping netfilter conntrack.
+		if option.Config.Tunnel != option.TunnelDisabled {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Fatalf("%s requires the agent to run in direct routing mode.", option.InstallNoConntrackIptRules)
```
nit: if we fatal here, why the preceeding `option.Config.InstallNoConntrackIptRules = false` or did you mean to just Warn instead of Fatal?

jibi replied:
Leftovers as before I was just emitting a warning, let me remove them :+1:

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1373,24 @@ func initEnv(cmd *cobra.Command) {
+		// the native routing CIDR, used to select all pod-to-pod
+		// traffic, can only be an IPv4 CIDR at the moment.
+		if !option.Config.EnableIPv4 {
+			option.Config.InstallNoConntrackIptRules = false
```
ditto

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -334,6 +334,17 @@ func initKubeProxyReplacementOptions() (strict bool) {
+		if !option.Config.KubeProxyReplacementFullyEnabled() {
+			option.Config.InstallNoConntrackIptRules = false
+			log.Fatalf("%s requires the agent to run with %s=%s.",
+				option.InstallNoConntrackIptRules, option.KubeProxyReplacement, option.KubeProxyReplacementStrict)
```
ditto

jibi committed:
iptables: add support for NOTRACK rules for pod-to-pod traffic

When running in KPR mode, all pod-to-pod traffic goes through connection
tracking twice: the first time traffic is tracked by netfilter and the
second one by the BPF datapath.

This is suboptimal as doing conntrack at the netfilter layer is not
strictly required for the operation of the CNI, but it has some
important performance implications.

This commit introduces a new agent option,
--install-no-conntrack-iptables-rules (disabled by default) which
installs `-j NOTRACK` Iptables rules that match all traffic from and to
the native-routing-cidr CIDR, to prevent netfilter from tracking all
pod-to-pod traffic.

This feature can be only enabled when Cilium is:

* running in direct routing
* running in full KPR mode
* not running its CNI chained on top of other CNI plugins
* not running in a managed Kubernetes environment such as AKS, EKS or GKE

There are a few reasons for these specific requirements:

* when running in tunneling mode (and with full KPR), the pod-to-pod
  traffic is encapsulated and decapsulated in the context of the BPF
  datapath program and so it is already skipping the netfilter raw table
  where connection tracking happens.
* when running in kube-proxy mode conntrack is required to perform NAT
* we cannot make assumptions on how a generic CNI is going to use
  netfilter conntrack

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
